### ISSUE TITLE
Identifiers not on both sides of the expression

### ISSUE THREAD

**Issue by jotaf98** [arogozhnikov/einops#29]
This seems like it should be (intuitively) plausible:

`rearrange(x, 'b -> a b c', a=1, c=1)`

to essentially push a vector to be compatible with some other tensors (for broadcasting operations). Currently this throws an error.

However, it's not possible. One (sort of ugly) workaround is:

`rearrange(x, '(a b c) -> a b c', a=1, c=1)`

However, it seems like this is a bit redundant and it obfuscates the intent a bit. Thoughts?

**Comment by arogozhnikov** [comment]
Thanks for asking. 

Use empty composition:

```python
y = rearrange(x, 'a -> () a () ')
# and back, it will verify that shape = [1, *, 1]
x = rearrange(y, '() a () -> a ')
```

I am not sure this is immediately obvious, but this is quite beatiful. Only axes of size one can be represented as empty composition (or decomposition). Those have specific meaning in `numpy` (and now pretty much everywhere), and those are a natural part of `einops` syntax. 

The following syntax I plan to use in one of following updates, but with another operation (but it should include your case of interest too).  
```
rearrange(x, 'b -> a b c', a=1, c=1)
```

**Comment by jotaf98** [comment]
That's a good suggestion, thanks! I'll be using it in the meanwhile.

I do like the last syntax though, it is very intuitive. May I ask what do you mean by "another operation"? Will there be a third one after rearrange and reduce?

**Comment by arogozhnikov** [comment]
> Will there be a third one after rearrange and reduce?

Yup. I haven't completely decide scope of this operation yet, but in minimal form it should cover `numpy.repeat` and `numpy.tile`.

**Comment by jotaf98** [comment]
Thanks. As a small follow-up, that shouldn't be too difficult, would it be possible to accept "1" as a short-hand for "()"? This could be implemented as a simple string replace.

I think this would have a high impact on readability (especially for people who don't know about the empty composition trick), without requiring any rewrite of core functionality (such as allowing numbers also for non-singleton dimensions, which has its own issues (discussed in #28).

**Comment by srama2512** [comment]
First of all, this library has made a huge difference in terms of writing readable and bug-free code. Thanks. Has there been any updates on adding the np.repeat / np.tile like operations? `torch.expand` has quite an ugly syntax at the moment.

**Comment by arogozhnikov** [comment]
@srama2512  thanks for pushing me. I'll try to find some spare time over following weekends to draft this feature

**Comment by arogozhnikov** [comment]
@srama2512 new repeat function is in the master. It is yet to appear in the documentation, but you're welcome to use it and share your feedback

**Comment by srama2512** [comment]
@arogozhnikov  - it works like a charm for my use cases. Einops feels complete!

**Comment by arogozhnikov** [comment]
@srama2512  awesome! Thanks for feedback

**Comment by jotaf98** [comment]
Thanks for being so responsive :)

May I ask what are your thoughts on this:

> As a small follow-up, that shouldn't be too difficult, would it be possible to accept "1" as a short-hand for "()"? This could be implemented as a simple string replace.

Not sure if you agree or not (hopefully yes!), but in any case, whatever the answer I think I can close the issue.

**Comment by arogozhnikov** [comment]
@jotaf98
 I'll just register here that I'm working on finding the right balance of convenience/safety/extensibility for this feature. I expect this to appear (at least in certain context) quite soon

**Comment by arogozhnikov** [comment]
@jotaf98 it is done and available in master branch. 

Axes of size 1 can be written as `1`, not `()`.

But also: reduced axes can be anonymous 
```python
reduce(image, 'h w 3 -> h w', 'mean')
```
as well as repeated axes
```python
repeat(image_bw, 'h w -> h w 3')
```

Same applies to subaxes - name is not required if those are reduced or repeated.

**Comment by jotaf98** [comment]
That's awesome, thanks! I'm sure that even people who have no familiarity with einops will intuitively understand this syntax.

Also, good idea on anonymous axes for reduce/repeat, there's no need to match them to any other axis so it's completely unambiguous!

**Comment by serycjon** [comment]
Going back to the original question, is there any problem in making: ```rearrange(x, 'b -> a b c', a=1, c=1)``` work? I know I can use ```()``` or ```1``` instead of ```a``` and ```c```, but I would still like to have the 1-dimensions named to make the code easy to understand.

**Comment by arogozhnikov** [comment]
@serycjon use repeat if you need to introduce new named axes. Rearrange is strict towards missing identifiers on any side by design
